### ISSUE TITLE
Adding a 2nd grouping field throws an exception

### ISSUE THREAD

**Issue by frantz45** [airbus-cyber/graylog-plugin-correlation-count#34]
Create an correlation count rule with 2 "group by fields".
When the rule triggers it throws the following exception:

```
2024-01-22T15:22:25.937+01:00 ERROR [EventProcessorEngine] Caught an unhandled exception while executing event processor <correlation-count/[Wizard] bbbb/65ae785be408b7238aeae2d8> - Make sure to modify the event processor to throw only EventProcessorException so we get more context!
java.lang.IndexOutOfBoundsException: index (1) must be less than size (1)
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1355) ~[graylog.jar:?]
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1337) ~[graylog.jar:?]
	at com.google.common.collect.SingletonImmutableList.get(SingletonImmutableList.java:44) ~[graylog.jar:?]
	at com.airbus_cyber_security.graylog.events.processor.correlation.CorrelationCountProcessor.associateGroupByFields(CorrelationCountProcessor.java:159) ~[?:?]
	at com.airbus_cyber_security.graylog.events.processor.correlation.CorrelationCountProcessor.runCheck(CorrelationCountProcessor.java:175) ~[?:?]
	at com.airbus_cyber_security.graylog.events.processor.correlation.CorrelationCountProcessor.createEvents(CorrelationCountProcessor.java:79) ~[?:?]
	at org.graylog.events.processor.EventProcessorEngine.execute(EventProcessorEngine.java:91) ~[graylog.jar:?]
	at org.graylog.events.processor.EventProcessorExecutionJob.execute(EventProcessorExecutionJob.java:115) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.executeJob(JobExecutionEngine.java:186) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.lambda$handleTrigger$2(JobExecutionEngine.java:159) ~[graylog.jar:?]
	at com.codahale.metrics.Timer.time(Timer.java:151) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.handleTrigger(JobExecutionEngine.java:159) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.lambda$execute$0(JobExecutionEngine.java:128) ~[graylog.jar:?]
	at org.graylog.scheduler.worker.JobWorkerPool.lambda$execute$0(JobWorkerPool.java:121) ~[graylog.jar:?]
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [graylog.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
2024-01-22T15:22:25.938+01:00 ERROR [EventProcessorExecutionJob] Event processor <correlation-count/65ae785be408b7238aeae2d8> failed to execute: Couldn't create events for: EventDefinitionDto{scope=DEFAULT, id=65ae785be408b7238aeae2d8, title=[Wizard] bbbb, description=, updatedAt=2024-01-22T14:21:14.440Z, priority=2, alert=true, config=CorrelationCountProcessorConfig{type=correlation-count, stream=65ae785be408b7238aeae2d0, additionalStream=65ae785be408b7238aeae2d3, additionalThresholdType=LESS, additionalThreshold=1, thresholdType=MORE, threshold=0, messagesOrder=ANY, searchWithinMs=60000, executeEveryMs=60000, groupingFields=[ddd, xxx], comment=Generated by the alert wizard, searchQuery=*}, fieldSpec={}, keySpec=[], notificationSettings=EventNotificationSettings{gracePeriodMs=0, backlogSize=100}, notifications=[Config{notificationId=65ae785be408b7238aeae2cd, notificationParameters=Optional.empty}], storage=[Config{type=persist-to-streams-v1, streams=[000000000000000000000002]}], schedulerCtx=null} (retry in 5000 ms)
org.graylog.events.processor.EventProcessorException: Couldn't create events for: EventDefinitionDto{scope=DEFAULT, id=65ae785be408b7238aeae2d8, title=[Wizard] bbbb, description=, updatedAt=2024-01-22T14:21:14.440Z, priority=2, alert=true, config=CorrelationCountProcessorConfig{type=correlation-count, stream=65ae785be408b7238aeae2d0, additionalStream=65ae785be408b7238aeae2d3, additionalThresholdType=LESS, additionalThreshold=1, thresholdType=MORE, threshold=0, messagesOrder=ANY, searchWithinMs=60000, executeEveryMs=60000, groupingFields=[ddd, xxx], comment=Generated by the alert wizard, searchQuery=*}, fieldSpec={}, keySpec=[], notificationSettings=EventNotificationSettings{gracePeriodMs=0, backlogSize=100}, notifications=[Config{notificationId=65ae785be408b7238aeae2cd, notificationParameters=Optional.empty}], storage=[Config{type=persist-to-streams-v1, streams=[000000000000000000000002]}], schedulerCtx=null}
	at org.graylog.events.processor.EventProcessorEngine.execute(EventProcessorEngine.java:105) ~[graylog.jar:?]
	at org.graylog.events.processor.EventProcessorExecutionJob.execute(EventProcessorExecutionJob.java:115) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.executeJob(JobExecutionEngine.java:186) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.lambda$handleTrigger$2(JobExecutionEngine.java:159) ~[graylog.jar:?]
	at com.codahale.metrics.Timer.time(Timer.java:151) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.handleTrigger(JobExecutionEngine.java:159) ~[graylog.jar:?]
	at org.graylog.scheduler.JobExecutionEngine.lambda$execute$0(JobExecutionEngine.java:128) ~[graylog.jar:?]
	at org.graylog.scheduler.worker.JobWorkerPool.lambda$execute$0(JobWorkerPool.java:121) ~[graylog.jar:?]
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [graylog.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [graylog.jar:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.lang.IndexOutOfBoundsException: index (1) must be less than size (1)
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1355) ~[graylog.jar:?]
	at com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1337) ~[graylog.jar:?]
	at com.google.common.collect.SingletonImmutableList.get(SingletonImmutableList.java:44) ~[graylog.jar:?]
	at com.airbus_cyber_security.graylog.events.processor.correlation.CorrelationCountProcessor.associateGroupByFields(CorrelationCountProcessor.java:159) ~[?:?]
	at com.airbus_cyber_security.graylog.events.processor.correlation.CorrelationCountProcessor.runCheck(CorrelationCountProcessor.java:175) ~[?:?]
	at com.airbus_cyber_security.graylog.events.processor.correlation.CorrelationCountProcessor.createEvents(CorrelationCountProcessor.java:79) ~[?:?]
	at org.graylog.events.processor.EventProcessorEngine.execute(EventProcessorEngine.java:91) ~[graylog.jar:?]
	... 12 more
```

**Comment by c8y3** [comment]
Reproduced with the previous configuration when sending several messages without field1 ({'version': '1.1', 'host': 'host', 'short_message': 'short_message', '_field2': '2', '_a': 'a', '_b': 'b'}).

**Comment by c8y3** [comment]
Reproduced with just an event definition of type correlation count:
* Default Stream, more than, 1
* Default Stream, more than, 1
* GroupBy fields "field1" "field2"
Then send 5 messages without field1 : {'version': '1.1', 'host': 'host', 'short_message': 'short_message', '_field2': '2'}

**Comment by c8y3** [comment]
More notes:
* strangely, does not fail if when only the first group by field (field1) is present.
* fails when no group by field is present
* It seems the graylog aggregation event has the same issue https://github.com/Graylog2/graylog2-server/issues/17314
* other related past issues: https://github.com/Graylog2/graylog2-server/pull/14678/commits/80311af3e94cc7b6bbc2c118bea081cb57d0b204 and https://github.com/Graylog2/graylog2-server/issues/14375
* the fact that, once the problem is encountered, Graylog repeats the same stack trace indefinitely and does not go one with the rest of the stream is worrying...

**Comment by c8y3** [comment]
In the different case scenarios, the key of the AggregationKeyResult is:
* when only field1 is present with value "value1", ["value1", "(Empty Value)"]
* when only field2 is present with value "value2", ["(Empty Value)"]
* when no group field is present, also ["(Empty Value)"]

**Comment by frantz45** [comment]
I think the 2nd solution is the best. I hope Graylog will fix the root cause

**Comment by c8y3** [comment]
Should be fixed in graylog-plugin-correlation-count 5.1.1. I still need to update the dependency in graylog-plugin-alert-wizard.
